In [1]:
import numpy as np
import pandas as pd

In [2]:
import scipy.io

In [3]:
data = scipy.io.loadmat('/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/arrhythmia.mat') 

In [4]:
X = data['X']
y = data['y']

In [5]:
from anomaly.base import AnomalyDetector

## Import our model

In [6]:
from models.xstream import xStream

In [15]:
pip install pysad

     |████████████████████████████████| 73 kB 4.2 MB/s eta 0:00:011
     |████████████████████████████████| 141 kB 22.6 MB/s eta 0:00:01
  Created wheel for pyod: filename=pyod-1.0.6-py3-none-any.whl size=175085 sha256=29b417165bce37ef2f26b5cc154e2e4633380243a64db998cb13f15226256794
  Stored in directory: /Users/Sophie/Library/Caches/pip/wheels/98/93/e6/6d40410d9635ecde42d06041a1ba7f2ee7396e036fcf702e73
Successfully built pyod
Note: you may need to restart the kernel to use updated packages.


In [17]:
pip install combo

  Created wheel for combo: filename=combo-0.1.3-py3-none-any.whl size=42871 sha256=149bdd0cd9c0f93168da1d2bab08d37a77a27d44c2e4aa3715b8a0f5485b0c55
  Stored in directory: /Users/Sophie/Library/Caches/pip/wheels/ec/44/39/0667fea44a2dfe692cc2a51f0f79ea49b9dee7def53594ef2e
Successfully built combo
Note: you may need to restart the kernel to use updated packages.


In [25]:
pip install mmh3

Note: you may need to restart the kernel to use updated packages.


In [7]:
from pysad.transform.probability_calibration import ConformalProbabilityCalibrator

In [8]:
from river.stream import iter_array

In [9]:
np.random.seed(61)

In [10]:
model = xStream()
calibrator = ConformalProbabilityCalibrator(windowed=True, window_size=300)
i=1
for xi, yi in iter_array(X,y):
    i += 1
    x_array = np.array(list(xi.values()))
    y_array = np.array(list(yi.values()))
    anomaly_score = model.score_one(x_array)
    model = model.learn_one(x_array)
#    anomaly_score = model.learn_score_one(x_array)
    
    calibrated_score = calibrator.fit_transform(anomaly_score)  # Fit & calibrate score.

    # Output if the instance is anomalous.
    if calibrated_score > 0.95:  # If probability of being normal is less than 5%.
        print(f"Alert: {i} data point is anomalous.")
#        print(xi)

Alert: 26 data point is anomalous.
Alert: 51 data point is anomalous.
Alert: 52 data point is anomalous.
Alert: 53 data point is anomalous.
Alert: 56 data point is anomalous.
Alert: 59 data point is anomalous.
Alert: 63 data point is anomalous.
Alert: 64 data point is anomalous.
Alert: 67 data point is anomalous.
Alert: 74 data point is anomalous.
Alert: 83 data point is anomalous.
Alert: 136 data point is anomalous.
Alert: 150 data point is anomalous.
Alert: 202 data point is anomalous.
Alert: 226 data point is anomalous.
Alert: 229 data point is anomalous.
Alert: 230 data point is anomalous.
Alert: 232 data point is anomalous.
Alert: 234 data point is anomalous.
Alert: 235 data point is anomalous.
Alert: 238 data point is anomalous.
Alert: 240 data point is anomalous.
Alert: 242 data point is anomalous.
Alert: 246 data point is anomalous.
Alert: 260 data point is anomalous.
Alert: 265 data point is anomalous.
Alert: 266 data point is anomalous.
Alert: 267 data point is anomalous.
Ale

score then learn (River):
Alert: 26 data point is anomalous.
Alert: 52 data point is anomalous.
Alert: 53 data point is anomalous.
Alert: 54 data point is anomalous.
Alert: 59 data point is anomalous.
Alert: 63 data point is anomalous.
Alert: 64 data point is anomalous.
Alert: 67 data point is anomalous.
Alert: 84 data point is anomalous.
Alert: 95 data point is anomalous.
Alert: 136 data point is anomalous.
Alert: 150 data point is anomalous.
Alert: 151 data point is anomalous.
Alert: 202 data point is anomalous.
Alert: 205 data point is anomalous.
Alert: 211 data point is anomalous.
Alert: 217 data point is anomalous.
Alert: 222 data point is anomalous.
Alert: 226 data point is anomalous.
Alert: 228 data point is anomalous.
Alert: 229 data point is anomalous.
Alert: 230 data point is anomalous.
Alert: 232 data point is anomalous.
Alert: 235 data point is anomalous.
Alert: 238 data point is anomalous.
Alert: 240 data point is anomalous.
Alert: 242 data point is anomalous.
Alert: 244 data point is anomalous.
Alert: 246 data point is anomalous.
Alert: 264 data point is anomalous.
Alert: 266 data point is anomalous.
Alert: 267 data point is anomalous.
Alert: 288 data point is anomalous.
Alert: 289 data point is anomalous.
Alert: 292 data point is anomalous.
Alert: 327 data point is anomalous.
Alert: 328 data point is anomalous.
Alert: 335 data point is anomalous.
Alert: 339 data point is anomalous.
Alert: 341 data point is anomalous.
Alert: 342 data point is anomalous.
Alert: 343 data point is anomalous.
Alert: 344 data point is anomalous.
Alert: 345 data point is anomalous.
Alert: 348 data point is anomalous.
Alert: 351 data point is anomalous.

learn then score (PySAD):
Alert: 24 data point is anomalous.
Alert: 51 data point is anomalous.
Alert: 52 data point is anomalous.
Alert: 53 data point is anomalous.
Alert: 67 data point is anomalous.
Alert: 76 data point is anomalous.
Alert: 83 data point is anomalous.
Alert: 126 data point is anomalous.
Alert: 136 data point is anomalous.
Alert: 150 data point is anomalous.
Alert: 205 data point is anomalous.
Alert: 211 data point is anomalous.
Alert: 217 data point is anomalous.
Alert: 222 data point is anomalous.
Alert: 226 data point is anomalous.
Alert: 228 data point is anomalous.
Alert: 229 data point is anomalous.
Alert: 232 data point is anomalous.
Alert: 234 data point is anomalous.
Alert: 235 data point is anomalous.
Alert: 238 data point is anomalous.
Alert: 240 data point is anomalous.
Alert: 242 data point is anomalous.
Alert: 246 data point is anomalous.
Alert: 249 data point is anomalous.
Alert: 266 data point is anomalous.
Alert: 276 data point is anomalous.
Alert: 326 data point is anomalous.
Alert: 327 data point is anomalous.
Alert: 328 data point is anomalous.
Alert: 330 data point is anomalous.
Alert: 333 data point is anomalous.
Alert: 334 data point is anomalous.
Alert: 335 data point is anomalous.
Alert: 338 data point is anomalous.
Alert: 339 data point is anomalous.
Alert: 340 data point is anomalous.
Alert: 341 data point is anomalous.
Alert: 342 data point is anomalous.
Alert: 343 data point is anomalous.
Alert: 344 data point is anomalous.
Alert: 347 data point is anomalous.
Alert: 348 data point is anomalous.

## PySAD version

In [11]:
from pysad.models import xStream
from pysad.transform.probability_calibration import ConformalProbabilityCalibrator
from pysad.utils import Data
import numpy as np

In [13]:
np.random.seed(61)

In [14]:
model = xStream()  # Init model.
calibrator = ConformalProbabilityCalibrator(windowed=True, window_size=300)  # Init probability calibrator.
streaming_data = Data().get_iterator("/Users/Sophie/Library/CloudStorage/OneDrive-Personnel/IP/Data Stream Processing/PySAD_to_River/data/arrhythmia.mat")  # Get streamer.

for i, (x, y_true) in enumerate(streaming_data):  # Stream data.
    anomaly_score = model.fit_score_partial(x)  # Fit to an instance x and score it.

    calibrated_score = calibrator.fit_transform(anomaly_score)  # Fit & calibrate score.

    # Output if the instance is anomalous.
    if calibrated_score > 0.95:  # If probability of being normal is less than 5%.
        print(f"Alert: {i}th data point is anomalous.")

Alert: 21th data point is anomalous.
Alert: 23th data point is anomalous.
Alert: 49th data point is anomalous.
Alert: 50th data point is anomalous.
Alert: 53th data point is anomalous.
Alert: 62th data point is anomalous.
Alert: 64th data point is anomalous.
Alert: 71th data point is anomalous.
Alert: 80th data point is anomalous.
Alert: 83th data point is anomalous.
Alert: 87th data point is anomalous.
Alert: 96th data point is anomalous.
Alert: 100th data point is anomalous.
Alert: 109th data point is anomalous.
Alert: 132th data point is anomalous.
Alert: 141th data point is anomalous.
Alert: 261th data point is anomalous.
Alert: 265th data point is anomalous.
Alert: 274th data point is anomalous.
Alert: 275th data point is anomalous.
Alert: 278th data point is anomalous.
Alert: 281th data point is anomalous.
Alert: 286th data point is anomalous.
Alert: 289th data point is anomalous.
Alert: 297th data point is anomalous.
Alert: 298th data point is anomalous.
Alert: 342th data point 